### PR TITLE
Create Cooldown contract

### PR DESCRIPTION
**Pull request by to-sh-1**
- Added cooldown contract and created claim for cooldown
- Added tests for TOKE related functions
- Updated graphio diagram

### CODE REVIEW

File: src/contracts/StakingCooldown.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.8.11;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract StakingCooldown {
```
Comment (0xean) [inline src/contracts/StakingCooldown.sol]: Isn't this identical to the warm up contract? if so, any reason why we cannot re-use the same contract (and maybe rename it so its more generic).

File: src/contracts/FoxStaking.sol [hunk 2]
```diff
@@ -259,12 +262,27 @@ contract FoxStaking is Ownable {
...
+        @param _amount uint
+        @return bool
+     */
+    function requestWithdrawalFromTokemak(uint256 _amount)
+        internal
+        returns (bool)
+    {
+        ITokePool tokePoolContract = ITokePool(tokePool);
+        tokePoolContract.requestWithdrawal(_amount);
+        return true; // TODO: TOKE requestWithdrawal function doesn't return anything.  Need to check for proper event emitted
```
Comment (0xean) [inline src/contracts/FoxStaking.sol]: Any reason to return true here? It doesn't appear that we are using the return value.

File: src/contracts/FoxStaking.sol [hunk 3]
```diff
@@ -313,6 +329,16 @@ contract FoxStaking is Ownable {
...
     }
 
+    /**
+        @notice stake FOX to enter warmup
+        @param _amount uint
+        @return bool
+     */
+    function stake(uint256 _amount) external returns (bool) {
+        stake(_amount, msg.sender);
+        return true;
```
Comment (0xean) [inline src/contracts/FoxStaking.sol]: same comment re: the true return val

File: src/contracts/FoxStaking.sol [hunk 4]
```diff
@@ -354,12 +380,27 @@ contract FoxStaking is Ownable {
...
         }
+
         IERC20(FOXy).safeTransferFrom(msg.sender, address(this), _amount);
-        // TODO: withdraw from TOKE and create cooldown claim
+
+        Claim memory info = cooldownInfo[msg.sender];
+        require(!info.lock, "Deposits for account are locked");
+
+        cooldownInfo[msg.sender] = Claim({
+            amount: info.amount.add(_amount),
```
Comment (0xean) [inline src/contracts/FoxStaking.sol:392]: This is probably a larger issue to tackle but we should do the following 

1. remove all of the hardcoded imports and use libraries (this was just a vestige of the forked code) and I think they did it for etherscan verification purposes, but we should avoid it if possible.
2. Upgrade to solidity 0.8.4 (or more recent) which has safe math built in
3. get ride of all of the unneeded safemath.

File: src/contracts/FoxStaking.sol [hunk 5]
```diff
@@ -412,6 +453,18 @@ contract FoxStaking is Ownable {
         warmupContract = _address;
     }
 
+    /**
+        @notice sets the contract address for LP staking
+        @param _address address
+     */
+    function setCooldownContract(address _address) external onlyManager {
```
Comment (0xean) [inline src/contracts/FoxStaking.sol]: One option to simplify our deployment script would be just to have the FoxStaking.sol constructor create the cooldown and warmup contracts.  We can then remove all of these setters and mark the addresses as constants

File: test/foxStakingTest.ts [hunk 6]
```diff
@@ -6,16 +6,21 @@ import { StakingWarmup } from "../typechain-types/StakingWarmup";
...
   let FOXy: Foxy;
   let foxStaking: FoxStaking;
   let fox: Contract;
+  let tFOX: Contract;
   let stakingWarmup: StakingWarmup;
+  let stakingCooldown: StakingCooldown;
 
   const FOX_WHALE = "0xF152a54068c8eDDF5D537770985cA8c06ad78aBB";
   const FOX = "0xc770EEfAd204B5180dF6a14Ee197D99d808ee52d";
+  const tFOXAddress = "0x808D3E6b23516967ceAE4f17a5F9038383ED5311";
```
Comment (0xean) [inline test/foxStakingTest.ts]: nit: pick a const format and stick with it.  IE FOX_WHALE vs tFOXAddress (tFOX_ADDRESS)

File: test/foxStakingTest.ts [hunk 7]
```diff
@@ -209,4 +225,77 @@ describe("FoxStaking", function () {
...
+        foxStaking.address,
+        stakingAmount2
+      );
+      await foxStakingStaker2.unstake(stakingAmount2, false);
+
+      const requestedWithdrawals = await tFOX.requestedWithdrawals(
+        foxStakingStaker1.address
+      );
+      expect(requestedWithdrawals.amount.eq(stakingAmount2)).true; // TODO: fix once able to stack requestedWithdrawals
+      expect(requestedWithdrawals.minCycle.eq(167)).true;
```
Comment (0xean) [inline test/foxStakingTest.ts]: maybe add a comment on wait minCycle is and why 167 is the correct value.

**Review by 0xean — COMMENTED** [review]
@toshiSat - Good to see this moving along.  I left some comments that are applicable to this PR and then that are probably future issues to tackle. Let me know if you have questions.

File: src/contracts/StakingCooldown.sol (hunk 1 above)
Comment (to-sh-1) [reply src/contracts/StakingCooldown.sol]: I thought the instant unstake would require more, but actually the forfeit function would handle that, so I think we can just combine these.  I'll add that in this PR

File: src/contracts/FoxStaking.sol (hunk 2 above)
Comment (to-sh-1) [reply src/contracts/FoxStaking.sol]: No, I was going with the format of stake, I wanted to talk to you about that.  I am concerned that requestWithdrawal will silently fail, but I'll leave the comment to make sure I investigate that later.

File: src/contracts/FoxStaking.sol (hunk 4 above)
Comment (to-sh-1) [reply src/contracts/FoxStaking.sol:392]: I'll make a ticket for this and push it up in a separate PR today or tomorrow

File: src/contracts/FoxStaking.sol [hunk 8]
```diff
@@ -248,6 +252,12 @@ contract FoxStaking is Ownable {
         require(_TokePool != address(0));
         tokePool = _TokePool;
 
+        Vesting warmup = new Vesting(address(this), FOXy);
```
Comment (0xean) [inline src/contracts/FoxStaking.sol:255]: This is nice, seems way cleaner.

**Review by 0xean — APPROVED** [review]

**Comment by 0xean** [comment]
missed that you added the ABI to src control, we should avoid this.

**Pull request merged**